### PR TITLE
Add some visualization code.

### PR DESCRIPTION
**Pull request by HudsonHuang**
Hi Ivanov, 

Thanks for your great work! It's a elegent and powerful RNN with Attention network.
But for convenience of visualization, I added some code here for Tensorboard monitoring.
Hope it helps.

Best regards!

### CODE REVIEW

**Comment by ilivans** [comment]
Hi @HudsonHuang ,

I appreciate your contribution. Let's try to improve the code before merging if you don't mind.

File: .gitignore [hunk 1]
```diff
@@ -87,3 +87,7 @@ ENV/
 
 # Rope project settings
 .ropeproject
+
+# Tensorboard logs
+.LENOVO-YOGA2PRO
```
Comment (ilivans) [inline .gitignore]: Your specific local directory shouldn't be here

File: train.py [hunk 2]
```diff
@@ -37,51 +38,74 @@
...
-
-# Embedding layer
-embeddings_var = tf.Variable(tf.random_uniform([vocabulary_size, EMBEDDING_DIM], -1.0, 1.0), trainable=True)
-batch_embedded = tf.nn.embedding_lookup(embeddings_var, batch_ph)
-
-# (Bi-)RNN layer(-s)
-rnn_outputs, _ = bi_rnn(GRUCell(HIDDEN_SIZE), GRUCell(HIDDEN_SIZE),
-                        inputs=batch_embedded, sequence_length=seq_len_ph, dtype=tf.float32)
-# rnn_outputs, _ = rnn(GRUCell(hidden_size), inputs=batch_embedded, sequence_length=seq_len_ph, dtype=tf.float32)
+batch_ph = tf.placeholder(tf.int32, [None, SEQUENCE_LENGTH],name='batch_ph')
```
Comment (ilivans) [inline train.py]: Missing whitespace after ','

File: train.py [hunk 3]
```diff
@@ -37,51 +38,74 @@
...
+target_ph = tf.placeholder(tf.float32, [None],name='target_ph')
+seq_len_ph = tf.placeholder(tf.int32, [None],name='seq_len_ph')
+keep_prob_ph = tf.placeholder(tf.float32,name='keep_prob_ph')
+
+with tf.name_scope('Encoder'):
```
Comment (ilivans) [inline train.py]: I don't actually understand where you find encoder and decoder in this architecture. It doesn't work like encoder-decoder, because we don't generate any sequence, we just solve a classification task. Tell me if I'm wrong.

File: train.py [hunk 4]
```diff
@@ -37,51 +38,74 @@
...
 seq_len_test = np.array([list(x).index(0) + 1 for x in X_test])
 seq_len_train = np.array([list(x).index(0) + 1 for x in X_train])
 
+merged = tf.summary.merge_all()
+
 # Batch generators
 train_batch_generator = batch_generator(X_train, y_train, BATCH_SIZE)
 test_batch_generator = batch_generator(X_test, y_test, BATCH_SIZE)
 
+train_writer = tf.summary.FileWriter('./logdir/train',accuracy.graph)
```
Comment (ilivans) [inline train.py]: Missing whitespace after ','

File: train.py [hunk 5]
```diff
@@ -37,51 +38,74 @@
...
+test_writer = tf.summary.FileWriter( './logdir/test',accuracy.graph)
+    
+
+session_conf = tf.ConfigProto(
```
Comment (ilivans) [inline train.py]: This statement fits one line I guess

File: train.py [hunk 6]
```diff
@@ -94,34 +118,41 @@
 
             # Training
             num_batches = X_train.shape[0] // BATCH_SIZE
-            for b in range(num_batches):
+            for b in tqdm(range(num_batches)):
                 x_batch, y_batch = next(train_batch_generator)
                 seq_len = np.array([list(x).index(0) + 1 for x in x_batch])  # actual lengths of sequences
-                loss_tr, acc, _ = sess.run([loss, accuracy, optimizer],
+                loss_tr, acc, _ ,summary= sess.run([loss, accuracy, optimizer,merged],
```
Comment (ilivans) [inline train.py]: Missing whitespace after ','

File: train.py [hunk 7]
```diff
@@ -94,34 +118,41 @@
...
-                accuracy_test += acc
-                loss_test += loss_test_batch
-            accuracy_test /= num_batches
-            loss_test /= num_batches
-
-            print("loss: {:.3f}, val_loss: {:.3f}, acc: {:.3f}, val_acc: {:.3f}".format(
-                loss_train, loss_test, accuracy_train, accuracy_test
-            ))
-        saver.save(sess, "model")
+            if epoch%10 == 0:
```
Comment (ilivans) [inline train.py]: Surround binary operators with a single space on either side

File: train.py [hunk 8]
```diff
@@ -94,34 +118,41 @@
...
+                print ("Testing")
```
Comment (ilivans) [inline train.py]: Extra space between `print` and parentheses

File: visualize.py [hunk 9]
```diff
@@ -10,7 +10,7 @@
 
 # Calculate alpha coefficients for the first test example
 with tf.Session() as sess:
-    saver.restore(sess, "model")
+    saver.restore(sess, "trained.model")
```
Comment (ilivans) [inline visualize.py]: Why did you change the name? And if it matters, shouldn't the name from line 9 be changed too? I would remain the old name for simplicity

File: train.py (hunk 7 above)
Comment (ilivans) [inline train.py]: There are just 3 epochs now. So this condition makes no sense. If test subset is small, we can perform evaluation each epoch, there is no problem.

File: train.py [hunk 10]
```diff
@@ -94,34 +118,41 @@
...
+                num_batches = X_test.shape[0] // BATCH_SIZE
+                for b in tqdm(range(num_batches)):
+                    x_batch, y_batch = next(test_batch_generator)
+                    seq_len = np.array([list(x).index(0) + 1 for x in x_batch])  # actual lengths of sequences
+                    loss_test_batch, acc,summary = sess.run([loss, accuracy,merged],
+                                                    feed_dict={batch_ph: x_batch,
```
Comment (ilivans) [inline train.py]: Bad indent

**Review by ilivans — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 1 above)
Comment (HudsonHuang) [reply .gitignore]: Yes, I forgot to delete this line, thank you for your review.

File: train.py (hunk 3 above)
Comment (HudsonHuang) [reply train.py]: I support that this was actually not a typical Encoder-Decoder architecture, but for the clarity of tensorboard graph, I think we should warp these ops in a namescope. How do you think about it?

File: train.py (hunk 7 above)
Comment (HudsonHuang) [reply train.py]: Yes, sorry for this line, I was running on my very slowly CPU notebook and wanted to speed up the progress, It was too late after I ran over all the code and find out there are only 3 epochs.

File: visualize.py (hunk 9 above)
Comment (HudsonHuang) [reply visualize.py]: Actually it Raised a Error on the original code:
```
Traceback (most recent call last):
  File "train.py", line 154, in <module>
    saver.save(sess, "model")
  File "D:\ProgramData\Anaconda3\lib\site-packages\tensorflow\python\training\saver.py", line 1594, in save
    raise exc
ValueError: Parent directory of model doesn't exist, can't save.
```
While I run on Windows with MINGW64, maybe there doesn't exist a avaiable saving path.
I tried this line may solve it:
```
saver.save(sess, "./model")
```

**Comment by HudsonHuang** [comment]
Thank you so much for your review, and sorry my careless, I learn alot from your code and this review.
Again, thanks for sharing your nice work with me.
Best wishes!

File: train.py (hunk 3 above)
Comment (ilivans) [reply train.py]: I think the 'decoder' scope will definitely confuse a lot of people, so it shouldn't be here. Wrap two scopes (embedding and rnn) into 'encoder' doesn't seem necessary for me, but this scope may stay if you find it helpful

File: visualize.py (hunk 9 above)
Comment (ilivans) [reply visualize.py]: Wow, it's so surprisingly. You say  `saver.save(sess, "model")` doesn't work meanwhile `saver.save(sess, "trained.model")` works? Let's make it `saver.save(sess, "./model")` then if it works for Windows and you don't mind.

**Review by ilivans — CHANGES_REQUESTED** [review]

**Comment by HudsonHuang** [comment]
Sorry for I did’t test through these code properly, and I think we fixed every bugs in this PR, right? 
Finally, I  appreciate your patience, and Lets close this PR now.

**Comment by ilivans** [comment]
I didn't run it yet, actually. I just wanted you to make cosmetic fixes first, hoping functionality is ok. The next step would be smoke test which will run smoothly I hope. Why did you close PR? =) We didn't finish yet and I really wanted to merge it soon.

**Comment by HudsonHuang** [comment]
That's embarrassing... OK, Ivanov, let me fix all these bugs.
Actually I have fixed them in my REPO, but I think I need more tests, I would make it perfect this time.

**Comment by ilivans** [comment]
No problem, I'm looking forward to hearing from you.

File: attention.py [hunk 11]
```diff
@@ -62,12 +62,14 @@ def attention(inputs, attention_size, time_major=False, return_alphas=False):
     b_omega = tf.Variable(tf.random_normal([attention_size], stddev=0.1))
     u_omega = tf.Variable(tf.random_normal([attention_size], stddev=0.1))
 
-    # Applying fully connected layer with non-linear activation to each of the B*T timestamps;
-    #  the shape of `v` is (B,T,D)*(D,A)=(B,T,A), where A=attention_size
-    v = tf.tanh(tf.tensordot(inputs, W_omega, axes=1) + b_omega)
+    with tf.name_scope('v'):
```
Comment (HudsonHuang) [inline attention.py:65]: Here does more than one operations, I think for better understanding how to produce 'v', It's  reasonable to warp these ops.

File: attention.py [hunk 12]
```diff
@@ -62,12 +62,14 @@ def attention(inputs, attention_size, time_major=False, return_alphas=False):
...
+        # Applying fully connected layer with non-linear activation to each of the B*T timestamps;
+        #  the shape of `v` is (B,T,D)*(D,A)=(B,T,A), where A=attention_size
+        v = tf.tanh(tf.tensordot(inputs, W_omega, axes=1) + b_omega)
+        
     # For each of the timestamps its vector of size A from `v` is reduced with `u` vector
-    vu = tf.tensordot(v, u_omega, axes=1)   # (B,T) shape
-    alphas = tf.nn.softmax(vu)              # (B,T) shape also
+    vu = tf.tensordot(v, u_omega, axes=1, name='vu')   # (B,T) shape
```
Comment (HudsonHuang) [inline attention.py]: It seems better in tensorboard if we give these ops a 'name'.

File: train.py [hunk 13]
```diff
@@ -37,51 +39,74 @@
 X_test = zero_pad(X_test, SEQUENCE_LENGTH)
 
 # Different placeholders
-batch_ph = tf.placeholder(tf.int32, [None, SEQUENCE_LENGTH])
-target_ph = tf.placeholder(tf.float32, [None])
-seq_len_ph = tf.placeholder(tf.int32, [None])
-keep_prob_ph = tf.placeholder(tf.float32)
+with tf.name_scope('inputs'):
```
Comment (HudsonHuang) [inline train.py]: It seems more clear to warp all input_phs into a namescope.

File: train.py [hunk 14]
```diff
@@ -37,51 +39,74 @@
...
 # (Bi-)RNN layer(-s)
 rnn_outputs, _ = bi_rnn(GRUCell(HIDDEN_SIZE), GRUCell(HIDDEN_SIZE),
                         inputs=batch_embedded, sequence_length=seq_len_ph, dtype=tf.float32)
+tf.summary.histogram('rnn_outputs', rnn_outputs)   
 # rnn_outputs, _ = rnn(GRUCell(hidden_size), inputs=batch_embedded, sequence_length=seq_len_ph, dtype=tf.float32)
 
+
 # Attention layer
-attention_output, alphas = attention(rnn_outputs, ATTENTION_SIZE, return_alphas=True)
+with tf.name_scope('Attention_layer'):
```
Comment (HudsonHuang) [inline train.py]: These histograms show how these parameters convergenced, but you can remove some of it if some seems verbose.

File: train.py [hunk 15]
```diff
@@ -37,51 +39,74 @@
...
 seq_len_test = np.array([list(x).index(0) + 1 for x in X_test])
 seq_len_train = np.array([list(x).index(0) + 1 for x in X_train])
 
+merged = tf.summary.merge_all()
+
 # Batch generators
 train_batch_generator = batch_generator(X_train, y_train, BATCH_SIZE)
 test_batch_generator = batch_generator(X_test, y_test, BATCH_SIZE)
 
+train_writer = tf.summary.FileWriter('./logdir/train', accuracy.graph)
```
Comment (HudsonHuang) [inline train.py:91]: Fixed whitespace

File: utils.py [hunk 16]
```diff
@@ -42,5 +42,5 @@ def batch_generator(X, y, batch_size):
     # Test batch generator
     gen = batch_generator(np.array(['a', 'b', 'c', 'd']), np.array([1, 2, 3, 4]), 2)
     for _ in range(8):
-        xx, yy = gen.next()
+        xx, yy = next(gen)
```
Comment (HudsonHuang) [inline utils.py:46]: Fixed batch_generator test.

File: visualize.py [hunk 17]
```diff
@@ -6,11 +6,11 @@
 """
 from train import *
 
-saver = tf.train.import_meta_graph('model.meta')
+saver = tf.train.Saver()
```
Comment (HudsonHuang) [inline visualize.py:9]: By using default saver, I found it could avoid Error, although I do not know the reason.

**Review by HudsonHuang — COMMENTED** [review]
Please checkout this commits, I have ran all code without error.

**Comment by ilivans** [comment]
image
Check by PEP8, please. Violations from `visualize.py` are mine, so fix them or not - at your discretion.

File: visualization.html [hunk 18]
```diff
@@ -1,71 +1,71 @@
...
-<font style="background: rgba(255, 255, 0, 0.059567)">both</font>
-<font style="background: rgba(255, 255, 0, 0.080328)">man</font>
-<font style="background: rgba(255, 255, 0, 0.412103)">and</font>
-<font style="background: rgba(255, 255, 0, 0.186147)">ape</font>
-<font style="background: rgba(255, 255, 0, 0.312056)">was</font>
-<font style="background: rgba(255, 255, 0, 0.542174)">outstanding</font>
-<font style="background: rgba(255, 255, 0, 0.518449)">not</font>
-<font style="background: rgba(255, 255, 0, 0.282087)">to</font>
-<font style="background: rgba(255, 255, 0, 0.431630)">mention</font>
+<font style="background: rgba(255, 255, 0, 0.181236)">how</font>
```
Comment (ilivans) [inline visualization.html]: You don't need to change the demonstration file.

File: visualize.py [hunk 19]
```diff
@@ -6,11 +6,11 @@
...
 
 # Calculate alpha coefficients for the first test example
 with tf.Session() as sess:
-    saver.restore(sess, "model")
+    saver.restore(sess, "./model")
```
Comment (ilivans) [inline visualize.py]: It differs from the path of the model in `train.py`, so the script doesn't work.
It will be nice to create global variable `MODEL_PATH` in `train.py` and import here.

File: train.py [hunk 20]
```diff
@@ -94,34 +119,39 @@
...
             loss_test /= num_batches
-
+    
             print("loss: {:.3f}, val_loss: {:.3f}, acc: {:.3f}, val_acc: {:.3f}".format(
                 loss_train, loss_test, accuracy_train, accuracy_test
             ))
-        saver.save(sess, "model")
+        train_writer.close()
+        test_writer.close()
+        saver.save(sess, "./logdir/model")
```
Comment (ilivans) [inline train.py]: I see 2 problems here:
- the name of the directory "logdir" doesn't correlate with its content (the model)
- if someone decides to disable tensorboard logging, the script will probably fail here because of the directory absence

The solution is to return "./model" path

File: train.py [hunk 21]
```diff
@@ -37,51 +39,74 @@
...
+merged = tf.summary.merge_all()
+
 # Batch generators
 train_batch_generator = batch_generator(X_train, y_train, BATCH_SIZE)
 test_batch_generator = batch_generator(X_test, y_test, BATCH_SIZE)
 
+train_writer = tf.summary.FileWriter('./logdir/train', accuracy.graph)
+test_writer = tf.summary.FileWriter( './logdir/test', accuracy.graph)
+    
+session_conf = tf.ConfigProto(
```
Comment (ilivans) [inline train.py]: Make it just
`session_conf = tf.ConfigProto(gpu_options=tf.GPUOptions(allow_growth=True))`
please

File: train.py [hunk 22]
```diff
@@ -94,34 +119,39 @@
...
-                                                      seq_len_ph: seq_len,
-                                                      keep_prob_ph: KEEP_PROB})
+                loss_tr, acc, _ , summary= sess.run([loss, accuracy, optimizer, merged],
+                                                    feed_dict={batch_ph: x_batch,
+                                                    target_ph: y_batch,
+                                                    seq_len_ph: seq_len,
+                                                    keep_prob_ph: KEEP_PROB})
                 accuracy_train += acc
                 loss_train = loss_tr * DELTA + loss_train * (1 - DELTA)
+                train_writer.add_summary(summary, b)
```
Comment (ilivans) [inline train.py]: Probably it should be `b + num_batches * epoch` instead of `b` here (for `test_writer` too) for correct continuous curves on tensorboard

File: train.py [hunk 23]
```diff
@@ -37,51 +39,74 @@
...
+    loss = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(logits=y_hat, labels=target_ph))
+    tf.summary.scalar('loss', loss)
+    optimizer = tf.train.AdamOptimizer(learning_rate=1e-3).minimize(loss)
+    
+    # Accuracy metric
+    accuracy = tf.reduce_mean(tf.cast(tf.equal(tf.round(tf.sigmoid(y_hat)), target_ph), tf.float32))
+    tf.summary.scalar('accuracy', accuracy)
 
 # Actual lengths of sequences
 seq_len_test = np.array([list(x).index(0) + 1 for x in X_test])
```
Comment (ilivans) [inline train.py]: These are unused variables, just noticed. Remove them, please

File: train.py (hunk 13 above)
Comment (ilivans) [reply train.py]: OK. It seems it would be better if scopes names will have the same style (starting with capital letter if it's a word, for example)

File: attention.py [hunk 24]
```diff
@@ -62,12 +62,14 @@ def attention(inputs, attention_size, time_major=False, return_alphas=False):
...
+    alphas = tf.nn.softmax(vu, name='alpha')              # (B,T) shape also
```
Comment (ilivans) [inline attention.py]: Let the scope name match the variable name not to be confused

File: train.py [hunk 25]
```diff
@@ -37,51 +39,74 @@
...
+with tf.name_scope('Embedding_layer'):
+    embeddings_var = tf.Variable(tf.random_uniform([vocabulary_size, EMBEDDING_DIM], -1.0, 1.0), trainable=True)
+    tf.summary.histogram('embeddings_var', embeddings_var)   
+    batch_embedded = tf.nn.embedding_lookup(embeddings_var, batch_ph)
 
 # (Bi-)RNN layer(-s)
 rnn_outputs, _ = bi_rnn(GRUCell(HIDDEN_SIZE), GRUCell(HIDDEN_SIZE),
                         inputs=batch_embedded, sequence_length=seq_len_ph, dtype=tf.float32)
+tf.summary.histogram('rnn_outputs', rnn_outputs)   
 # rnn_outputs, _ = rnn(GRUCell(hidden_size), inputs=batch_embedded, sequence_length=seq_len_ph, dtype=tf.float32)
```
Comment (ilivans) [inline train.py]: You can remove this dead line too as you removed unused import of `rnn`

**Review by ilivans — CHANGES_REQUESTED** [review]

File: train.py [hunk 26]
```diff
@@ -8,13 +8,15 @@
 from __future__ import print_function, division
 
 import tensorflow as tf
+import numpy as np
```
Comment (ilivans) [inline train.py]: Do `Code -> Optimize imports` in PyCharm (if you use it) to sort imports lexicographically

**Comment by ilivans** [comment]
Final comments I hope. With one bug fixed all works for me too.

File: visualize.py (hunk 19 above)
Comment (HudsonHuang) [reply visualize.py]: Brilliant solution.

File: train.py (hunk 23 above)
Comment (HudsonHuang) [reply train.py]: I don't quite understand, did you means removing the variable 'accuracy'? I guess accuracy is necessary in training progress, especially when someone decides to disable tensorboard logging, he can still get an accuracy on training progress.

File: visualize.py [hunk 27]
```diff
@@ -36,3 +36,6 @@
         elif word == ":PAD:":
             break
         html_file.write('<font style="background: rgba(255, 255, 0, %f)">%s</font>\n' % (alpha, word))
+
+print("")
```
Comment (HudsonHuang) [inline visualize.py]: Add an ‘\n’ here, avoid messy notice .

File: visualize.py [hunk 28]
```diff
@@ -20,11 +20,11 @@
 
 # Build correct mapping from word to index and inverse
 word_index = imdb.get_word_index()
-word_index = {word:index + INDEX_FROM for word, index in word_index.items()}
+word_index = {word: index + INDEX_FROM for word, index in word_index.items()}
 word_index[":PAD:"] = 0
 word_index[":START:"] = 1
 word_index[":UNK:"] = 2
-index_word = {value:key for key,value in word_index.items()}
+index_word = {value: key for key, value in word_index.items()}
```
Comment (HudsonHuang) [inline visualize.py:27]: Fixed PEP8, PyCharm is a great tool, thank you for your suggestion!

**Review by HudsonHuang — COMMENTED** [review]
Only left the question of whether to delete those variables or not, we are almost done!  I think it would be better to support both ways for monitoring, if you don't mind.

File: train.py (hunk 23 above)
Comment (ilivans) [reply train.py]: No no, I mean `seq_len_test` and `seq_len_train`. Probably they were unused before PR, but nevertheless.

File: visualize.py (hunk 27 above)
Comment (ilivans) [reply visualize.py]: What do you mean by "messy notice"? Instead of this line you can just add "\n" to the next:
`print("\nOpen ...`

**Comment by ilivans** [comment]
Added 2 replies (hope you'll notice them). 
Also I didn't see you removed `visualization.html` from commits. I guess you added it accidentally.

File: train.py (hunk 23 above)
Comment (HudsonHuang) [reply train.py]: OK. I cleaned them, and I don't find any other unused variables there.

**Comment by HudsonHuang** [comment]
Seems nearly finished this time.

**Review by ilivans — APPROVED** [review]

**Comment by ilivans** [comment]
Great job! Thank you @HudsonHuang !

**Comment by HudsonHuang** [comment]
That's great! Thank you and the same to you.
